drltrace/drltrace_frontend.cpp

```diff
@@ -0,0 +1,314 @@
+#endif
+    void *inject_data;
+    int exitcode;
+    char **argv;
+    char *tmp;
+    const char *target_app_name;
+
+    char full_target_app_path[MAXIMUM_PATH];
+    char tmp_path[MAXIMUM_PATH];
+    char full_frontend_path[MAXIMUM_PATH];
+    char full_dr_root_path[MAXIMUM_PATH];
+    char full_drlibtrace_path[MAXIMUM_PATH];
+
+    int last_index;
+    std::string parse_err;
+    drfront_status_t sc;
+
+#if defined(WINDOWS) && !defined(_UNICODE)
+# error _UNICODE must be defined
+#else
+    /* Convert to UTF-8 if necessary */
+    sc = drfront_convert_args((const TCHAR **)targv, &argv, argc);
+    if (sc != DRFRONT_SUCCESS)
+        DRLTRACE_ERROR("failed to process args, error code = %d\n", sc);
+#endif
+
+    if (!droption_parser_t::parse_argv(DROPTION_SCOPE_FRONTEND, argc, (const char **)argv,
+                                       &parse_err, &last_index) || argc < 2)
+        DRLTRACE_ERROR("Usage error: %s\n Usage:\n%s\n", parse_err.c_str(),
+              droption_parser_t::usage_short(DROPTION_SCOPE_ALL).c_str());
+
+    if (op_version.get_value()) {
+        print_version();
+        return 0;
+    }
+    if (op_help.get_value()) {
+        printf("Usage:\n%s", droption_parser_t::usage_long(DROPTION_SCOPE_ALL).c_str());
+        return 0;
+    }
+
+    target_app_name = argv[last_index];
+    if (target_app_name == NULL) {
+        DRLTRACE_ERROR("Usage error, target application is not specified.\n Usage:\n%s\n",
+                       droption_parser_t::usage_short(DROPTION_SCOPE_ALL).c_str());
+    }
+    sc = drfront_get_app_full_path(target_app_name, full_target_app_path,
+                                   BUFFER_SIZE_ELEMENTS(full_target_app_path));
+    if (sc != DRFRONT_SUCCESS)
+        DRLTRACE_ERROR("drfront_get_app_full_path failed on %s, error code = %d\n",
+                       target_app_name, sc);
+
+    /* get DR's root directory and drltracelib.dll full path */
+    sc = drfront_get_app_full_path(argv[0], full_frontend_path,
+                                   BUFFER_SIZE_ELEMENTS(full_frontend_path));
+    if (sc != DRFRONT_SUCCESS)
+        DRLTRACE_ERROR("drfront_get_app_full_path failed on %s, error code = %d\n",
+                       argv[0], sc);
+
+    tmp = full_frontend_path + strlen(full_frontend_path) - 1;
+
+    while (*tmp != DIRSEP && *tmp != ALT_DIRSEP && tmp > full_frontend_path)
+        tmp--;
+    *(tmp+1) = '\0';
+    _snprintf(tmp_path, BUFFER_SIZE_ELEMENTS(tmp_path) +
+              BUFFER_SIZE_ELEMENTS("../dynamorio"),
+              "%s../dynamorio", full_frontend_path);
+
+    sc = drfront_get_absolute_path(tmp_path, full_dr_root_path,
+                                   BUFFER_SIZE_ELEMENTS(tmp_path));
+    if (sc != DRFRONT_SUCCESS)
+        DRLTRACE_ERROR("drfront_get_absolute_path failed, error code = %d\n", sc);
+
+    _snprintf(full_drlibtrace_path, BUFFER_SIZE_ELEMENTS(full_frontend_path) +
+              BUFFER_SIZE_ELEMENTS(drlibpath) + BUFFER_SIZE_ELEMENTS(".dll"),
```

Comment:
The max size should be the size of the destination buffer, unless you're truncating the string which does not seem to be the case